datafusion-cli/src/object_storage.rs

```diff
@@ -43,28 +46,61 @@ impl FromStr for ObjectStoreScheme {
     }
 }
 
-#[derive(Debug)]
-pub struct DatafusionCliObjectStoreProvider {}
+#[derive(Debug, Default)]
+pub struct DatafusionCliObjectStoreRegistry {
+    inner: DefaultObjectStoreRegistry,
+}
+
+impl DatafusionCliObjectStoreRegistry {
+    pub fn new() -> Self {
+        Default::default()
+    }
+}
+
+impl ObjectStoreRegistry for DatafusionCliObjectStoreRegistry {
+    fn register_store(
+        &self,
+        scheme: &str,
+        host: &str,
+        store: Arc<dyn ObjectStore>,
+    ) -> Option<Arc<dyn ObjectStore>> {
+        self.inner.register_store(scheme, host, store)
+    }
+
+    fn put_with_url(
+        &self,
+        url: &Url,
+        store: Arc<dyn ObjectStore>,
+    ) -> Option<Arc<dyn ObjectStore>> {
+        self.inner.put_with_url(url, store)
+    }
+
+    fn get_by_url(&self, url: &Url) -> Result<Arc<dyn ObjectStore>> {
+        self.inner.get_by_url(url).or_else(|_| {
+            let store =
+                ObjectStoreScheme::from_str(url.scheme()).map(
+                    |scheme| match scheme {
+                        ObjectStoreScheme::S3 => build_s3_object_store(url),
+                        ObjectStoreScheme::GCS => build_gcs_object_store(url),
+                    },
+                )??;
```

Comment:
```suggestion
            let store = match ObjectStoreScheme::from_str(url.scheme())?
                    ObjectStoreScheme::S3 => build_s3_object_store(url)?,
                    ObjectStoreScheme::GCS => build_gcs_object_store(url)?,
            };
```

Or something similar